### PR TITLE
Detailedlist gtk listbox

### PR DESCRIPTION
**Pull request by meow464**
Better implements gtk detailedlist. It uses Gtk.ListBox, includes a toggable delete button (and other buttons in the future), a refresh button and a scroll to bottom button.

The previous implementation used Gtk.TreeView which only supports simple contents such as labels and images. It did not have any buttons.

Note: When you right click a row it exposes the delete button. On touch screens this should happen with a long press. I don't have a touch screen so I can't test.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by meow464** [comment]
Note: When you right click a row it exposes the delete button. On touch screens this should happen with a long press. I don't have a touch screen so I can't test.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 1]
```diff
@@ -1,92 +1,185 @@
...
 
-    def set_on_delete(self, handler):
-        pass
+    def index(self, item: 'Row'):
+        for index in range(0, len(self.store)):
+            if item._impl == self.store[index]:
+                return index
+        return None
+
+    def _on_refresh(self):
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:144]: Convention is to name these methods `gtk_on_<gtk_signal_name>`;

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 2]
```diff
@@ -1,92 +1,185 @@
...
-    we should use a Gtk.ListBox (see gtk3-demo).
+    Gtk DetailedList implementation.
+    Gtk.ListBox inside a Gtk.ScrolledWindow.
     """
 
     def create(self):
-        self.renderer = IconTextRenderer.from_style(self.interface)
+        self._on_refresh_handler = None
+        self._on_select_handler = None
+        self._on_delete_handler = None
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:17]: What is the purpose of these variables? If there's a user-provided handler, it should be declared and provide by the interface layer (i.e., `self.interface.on_refresh_handler`)

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 3]
```diff
@@ -1,92 +1,185 @@
...
+        
+        # Not the same as selected row. _active_row is the one with its buttons exposed.
+        self._active_row = None
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:20]: Generally, we try to avoid having a Python-side proxy of "state", unless it's completely unavoidable. Is there no way to programatically extract the active row from the ListBox?

File: src/gtk/toga_gtk/app.py [hunk 4]
```diff
@@ -107,6 +107,16 @@ def gtk_startup(self, data=None):
         settings = Gtk.Settings.get_default()
         settings.set_property("gtk-shell-shows-menubar", False)
 
+        # Set any custom styles
+        css_provider = Gtk.CssProvider()
+        css_provider.load_from_data(STYLES)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/app.py:112]: I take it this is an "application wide" style sheet? If so, we may want to rename `STYLES` to reflect that.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 5]
```diff
@@ -1,92 +1,185 @@
...
 
-    def set_on_refresh(self, handler):
-        pass
+        self.native.interface = self.interface
 
-    def after_on_refresh(self):
-        # No special handling required
-        pass
+        self._on_select_signal_handler = self.list_box.connect(
+            'row-selected', lambda w, item: self._on_select(item))
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:53]: Given `_on-select` is an internal method, why not give `_on_select` the required prototype, rather than funnelling through a lambda?

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 6]
```diff
@@ -1,92 +1,185 @@
...
+
+    def change_source(self, source: 'ListSource'):
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We have to wait until the rows are actually added decide how to position the
+        # refresh button
+        GLib.idle_add(lambda: not self._changed())
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:83]: What does this do? It looks like it's adding a global run loop callback... which I would have thought would come with a performance hit.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Looks like a good first pass - nice work. 

I've added a couple of initial comments regarding some structural questions/concerns; there's also the flake8 test to deal with. The biggest problem I see so far is the way signal handlers are being managed; it looks like you're replacing (or ignoring?) interface-level handler management. 

On top of that, there's a couple of high level behaviors that seem a little off:

1. The refresh button seems like it's on the wrong end - when I hit the bottom of the list, the refresh appears at the *top* of the page.
2. It's not clear what the "down arrow with a dot" button is meant to do. It looks like it's a "scroll to bottom" button... but it's not clear to me (a) why that is needed, or (b) why there isn't a corresponding scroll to top button.
3. The Right-to-left text examples don't start at the right hand side of the page - they're right to left *text*, but the box is rendered in a LTR layout (i.e., it's on the far left, not the far right). 

(1) and (2) might be idiomatic things - they feel alien to *me*, but I'm not a daily GTK user,

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 2 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:17]: It's the user handler that is set by `set_on_refresh`, `set_on_delete` and `set_on_select`. Those functions have a `handler: callable` argument.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 3 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:20]: This is not the selected row (clicked and highlighted in blue), it's the row with the delete button showing. ListBox has no knowledge of this.
The purpose of this variable is hiding the delete button on one row when the user clicks somewhere else. The initial approach I used was a custom signal sent  to all rows `emit on_list_clciked(item: Gtk.ListBoxRow)` but this is not very elegant. If there were 150 rows, all of them would receive a signal, compare the argument to `self` and only one would do something.

File: src/gtk/toga_gtk/app.py (hunk 4 above)
Comment (meow464) [reply src/gtk/toga_gtk/app.py:112]: Yes, it is application wide. What do you think would be a good name?
Note: At the moment it only has one class that does not overwrite anything native to gtk.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 5 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:53]: `w` is `self.list_box`, this felt cleaner because `_on_select` doesn't need `w` but I'm not married to this style.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 6 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:83]: Display of the scroll and refresh buttons depends on how many rows there are on the list, or more specifically on the dimensions of the list. 

We can't know the dimensions of each row (and thus of the list) until gtk allocates space for it. Gtk does emit `size-allocate` after allocation but I couldn't find any guarantees that the rows have their sizes allocated in the order they are inserted in the `ListStore` and in my opinion that's unlikely to be the case. 

Therefore we would need to wait for `size-allocate` on all rows and either update the visibility of the buttons on all `size-allocates` or have a counter and only do it on the last `size-allocate`. Obviously none of those options are desirable.

Fortunately functions added with `idle_add` are run when gtk is idle and thus after any size allocation. This solves our problem and from the perspective of the user this happens immediately.

With respect to the performance hit: even though we are adding the callback to the global loop, it only runs once. This is what that lambda is for. If a callback returns `False` then it's not ran again. I used a lambda because returning `False` from `self._changed()` would mean returning `False` on success.

I'm pretty sure I had a comment explaining this. Guess I deleted it by accident? Should I add the text above as a comment?

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 1 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:144]: Will do.

**Comment by meow464** [comment]
>     1. The refresh button seems like it's on the wrong end - when I hit the bottom of the list, the refresh appears at the _top_ of the page.
The rationale is: When the user is at the bottom of the list, they are more likely to be interested in the content at the bottom of the list and thus the button appears at the top. Same when they are on the top of the list.
If the button appeared on the bottom it would also block the view of the bottom row.

This is not a idiomatic thing. I couldn't find any gtk application that had a generic refresh button. The closest I did find was Fractal that has infinite scroll at the top but this is not very generic.
If our user is building a chat application, infinite scroll on top is great but on the bottom it makes no sense. If they are building a "scroll feed", infinite scroll on bottom is great but might or might not make sense on top.

What Fractal does have is the hovering scroll to bottom button and that's where I drove inspiration from.
 
>     2. It's not clear what the "down arrow with a dot" button is meant to do. It looks like it's a "scroll to bottom" button... but it's not clear to me (a) why that is needed, or (b) why there isn't a corresponding scroll to top button.
That icon is a symbolic icon, it's called "go-bottom-symbolic". On my theme it's an arrow pointing down to a horizontal line, to be honest it looks like a download button but that's just how cross theme gtk is.
(a) Fractal (gnome's matrix client) has it and it's something users expect on a chat client.
(b) Scroll to top is weird in _chats_ but I'm not opposed to adding it.

>     3. The Right-to-left text examples don't start at the right hand side of the page - they're right to left _text_, but the box is rendered in a LTR layout (i.e., it's on the far left, not the far right).

I completely missed this.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 1 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py]: Will do.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 2 above)
Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/detailedlist.py:17]: Sure - but why are they being copied here? Why not access them as self.interface.on_refresh_handler?

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 7]
```diff
@@ -1,92 +1,185 @@
...
+        self.right_click_gesture = Gtk.GestureMultiPress.new(self.list_box)
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self._on_right_click)
+
+    @property
+    def on_delete(self):
+        if self._on_delete_handler is not None:
+            return self.gtk_on_delete_clicked
+        return None
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:64]: What is the purpose of this property? And why is it mixing GTK event handler with Toga event handlers?

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 6 above)
Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/detailedlist.py:83]: That would definitely be helpful. The combination of the lambda, the GTK handling of a `false` return value, and the slightly opaque name `_changed` means it's not at all clear that this will be invoked once.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 8]
```diff
@@ -1,92 +1,185 @@
...
 
-    def set_on_refresh(self, handler):
-        pass
+        self.native.interface = self.interface
 
-    def after_on_refresh(self):
-        # No special handling required
-        pass
+        self._on_select_signal_handler = self.list_box.connect(
+            'row-selected', lambda w, item: self.gtk_on_row_selected(item))
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:53]: The use of lambda here is excessive. We're the ones defining `gtk_on_row_selection` - so we can give it whatever prototype we want. If it needs an extra `widget` argument that will be ignored, so be it.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 9]
```diff
@@ -1,92 +1,185 @@
...
+
+        self.right_click_gesture = Gtk.GestureMultiPress.new(self.list_box)
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self._on_right_click)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:58]: Convention is that any system-specific message handler will have a `gtk_` prefix, so it can be clearly identified.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 10]
```diff
@@ -1,92 +1,185 @@
...
+
+    @property
+    def on_delete(self):
+        if self._on_delete_handler is not None:
+            return self.gtk_on_delete_clicked
+        return None
+
+    def row_factory(self, item: 'Row'):
+        new_item = TextIconRow(self.interface.factory, self, item)
+        setattr(item, '_impl', new_item)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:68]: Why not just `item._impl = new_item`?

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 11]
```diff
@@ -1,92 +1,185 @@
...
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self._on_select_signal_handler)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:72]: As above; GTK handler should have a `gtk_` prefix

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 12]
```diff
@@ -1,92 +1,185 @@
...
+    def set_on_refresh(self, handler: callable):
+        if handler is not None:
+            self._on_refresh_handler = handler
+            self.refresh_button.set_on_refresh(self.gtk_on_refresh_clicked)
 
-    def set_on_select(self, handler):
+    def set_on_select(self, handler: callable):
+        self._on_select_handler = handler
+        if handler is not None:
+            self.list_box.set_selection_mode(Gtk.SelectionMode.SINGLE)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:129]: The existence of a selection handler isn't what controls the selectability of rows; DetailedList rows should always be selectable.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 13]
```diff
@@ -1,92 +1,185 @@
...
+        self._on_delete_handler = handler
+
+    def after_on_refresh(self):
         # No special handling required
         pass
 
-    def set_on_delete(self, handler):
-        pass
+    def index(self, item: 'Row'):
+        for index in range(0, len(self.store)):
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:139]: It's not clear why the *widget* is doing this iteration. The source can give you the position of the item in the list (and also gives the option of optimising the lookup).

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 14]
```diff
@@ -1,92 +1,185 @@
...
+
+    def gtk_on_delete_clicked(self, item: Gtk.ListBoxRow):
+        if self._on_delete_handler is not None:
+            if self._active_row == item:
+                self._active_row = None
+
+            self.interface.data.remove(item.interface)
+            self._on_delete_handler(self.interface, item.interface)
+
+    def _changed(self):
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:164]: As noted above, the naming of this method is a little opaque. There are many things that can change in a widget; it would be helpful if the naming of the method gave some indication of the context of the change where it needs to be invoked (as well as a docstring to clarify further)

File: src/gtk/toga_gtk/widgets/internal/buttons/__init__.py [hunk 15]
```diff
@@ -0,0 +1,2 @@
+from .refresh import RefreshButton
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/__init__.py:1]: Why is this needed? It's an internal hierarchy; we can import directly from the source of definition.

File: src/gtk/toga_gtk/widgets/internal/buttons/base.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
+class ParentPosition:
+    def __init__(self, adj):
+        self.adj = adj
+
+    def _is_parent_scrollable(self):
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/base.py:5]: Why are all these methods marked as "private" (i.e., with an underscore prefix)? Effectively, this is a class with no public API - and "public" in this case is "the implementation layer", so there's already no risk of the end-user of Toga being exposed to the class.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+
+class RefreshButtonWidget(Gtk.Revealer):
+    def __init__(self, on_refresh: callable, on_close: callable, 
+                 position: Gtk.Align, margin: int, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.parent = None
+
+        self._refresh_btn = Gtk.Button.new_from_icon_name(
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:12]: As above - why the "private" modifier?

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+            "clicked",
+            lambda w: on_refresh())
+
+        self._close_btn = Gtk.Button.new_from_icon_name(
+            "close-symbolic", Gtk.IconSize.BUTTON)
+        self._close_btn.set_can_focus(False)
+        close_btn_context = self._close_btn.get_style_context()
+        close_btn_context.add_class("osd")
+        close_btn_context.add_class("toga-detailed-list-floating-buttons")
+        refresh_btn_context.remove_class("button")
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:29]: Is there a reason this is here, and not a few lines further up with all the other refresh button code?

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+                 position: Gtk.Align, margin: int, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.parent = None
+
+        self._refresh_btn = Gtk.Button.new_from_icon_name(
+            "view-refresh-symbolic", Gtk.IconSize.BUTTON)
+        self._refresh_btn.set_can_focus(False)
+        refresh_btn_context = self._refresh_btn.get_style_context()
+        refresh_btn_context.add_class("osd")
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:16]: "osd" isn't a very meaningful class name; is there an internal GTK reason for using this name, or could we pick something more verbose?

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+        refresh_btn_context.add_class("toga-detailed-list-floating-buttons")
+
+        self._refresh_btn_handler = self._refresh_btn.connect(
+            "clicked",
+            lambda w: on_refresh())
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:21]: As before - we're in control of the handler, so we can give it whatever prototype we want. In addition, since it's a GTK handler, we should use the `gtk_` prefix.

It's possible that this might also fall foul of binding the widget to a specific handler; the handlers can be changed at runtime; we should be making a permanent event binding to an internal method, and using *that* method to direct to the "current" `on_refresh()` handler, rather than binding directly to the `on_refresh` that was in place when the button was created.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+            "close-symbolic", Gtk.IconSize.BUTTON)
+        self._close_btn.set_can_focus(False)
+        close_btn_context = self._close_btn.get_style_context()
+        close_btn_context.add_class("osd")
+        close_btn_context.add_class("toga-detailed-list-floating-buttons")
+        refresh_btn_context.remove_class("button")
+
+        self._close_btn_handler = self._close_btn.connect(
+            "clicked", 
+            lambda w: on_close())
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:33]: Same comment here as with on_refresh.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+        
+        self.button_top = RefreshButtonWidget(self._on_refresh_clicked,
+                                              self._on_close_clicked,
+                                              Gtk.Align.START,
+                                              self.margin)
+
+        self.button_bottom = RefreshButtonWidget(self._on_refresh_clicked,
+                                                 self._on_close_clicked,
+                                                 Gtk.Align.END,
+                                                 self.margin)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:96]: Same comment re: `gtk_` prefixing.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+
+        self.adj_handler = self.adj.connect(
+            "value-changed",
+            lambda adj: self.list_changed())
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:100]: Same comment about `gtk_` prefixing and defining a full handler.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I've reviewed a big chunk of this, and it looks like it's on the right track; however, there's still a few oddities with handlers and naming, and it looks like these carry on through the rest of the code that I haven't reviewed.

There's also the flake8 failures picked up by CI; it would be worth running that locally before submitting the PR to pick up code style discrepancies.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 7 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:64]: `set_on_refresh` receives a callback, it felt like I should save it. I will use `self.interface.on_refresh_handler` then. 

The row class calls `on_delete` to know whether it should show a delete button or not. It doesn't show the button if a handler is not set. I shouldn't have renamed this `gtk_on_delete_clciked`.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 13 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:139]: Neither `Gtk.ListBox` nor `Gio.ListStore` have an `index` method so I assumed it simply wasn't available. Turns out `Gtk.ListBoxRow` has `get_index`. That loop was really upsetting me.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 12 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:129]: Ok.
What about the delete button? Should rows be always deleteable or only when `delete_handler` is set?

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py (hunk 19 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:16]: It's a Gtk class. We need to remove it for the refresh and scroll button to look the way they do. Otherwise they retain their 3d-ness.

File: src/gtk/toga_gtk/widgets/internal/buttons/base.py (hunk 16 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/base.py:5]: This is a base class for scrollable rows, it's meant to be inherited from. The inheriting class will only use these methods internally.

File: src/gtk/toga_gtk/widgets/internal/buttons/__init__.py (hunk 15 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/__init__.py:1]: `from internal.buttons import RefreshButton` feels cleaner than `from internal.buttons.refresh import RefreshButton`
`internal.buttons` also provides `ParentPosition` and `ScrollButton`.

Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/internal/buttons/__init__.py:1]: Cleaner by what measure? You're obscuring where `RefreshButton` is actually defined. This sort of optimisation makes sense when it's an external API and you want to provide a clean API surface (e.g., the `toga.*` namespace) - but for an explicitly internal API, it only serves to make the implementation harder to reason about. If the number of import paths is unwieldy, that perhaps indicates that the package hierarchy itself is the problem.

File: src/gtk/toga_gtk/widgets/internal/buttons/base.py (hunk 16 above)
Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/internal/buttons/base.py:5]: So... and internal API, that will be subclassed, and the methods *must* be used in the place where it's subclassed. Underscore prefixing is effectively a "you shouldn't be using this method" API - why are we using a prefix here when we explicitly *will* need to use the API?

File: src/gtk/toga_gtk/widgets/internal/buttons/__init__.py (hunk 15 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/__init__.py:1]: Makes sense.

File: src/gtk/toga_gtk/widgets/internal/buttons/base.py (hunk 16 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/base.py:5]: Makes sense.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py (hunk 20 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:21]: Added `gtk_` here and to the class below.

That is an internal method. The class below `RefreshButton` has a `set_on_refresh` method. This class `RefreshButtonWidget` only handles the refresh button presentation, since there will be two buttons it makes sense to move the code to its own class.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py (hunk 23 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:100]: `list_changed` can be called externally. `DetailedList` will call this method after, adding, removing or changing  a row.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py (hunk 22 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:96]: `gtk_on_refresh` and `gtk_on_close`  are now also set with a `set_on_refresh` and `set_on_close`. This increases consistency with the rest of the code.

File: src/gtk/toga_gtk/widgets/internal/buttons/refresh.py (hunk 23 above)
Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/internal/buttons/refresh.py:100]: If `list_changed()` is called externally, then it should continue to exist as-is. What I'm saying is explicit is better than implicit: having both `list_changed()` (the public method) and `gtk_value_changed()` (an explicit handler whose implementation calls `list_changed()` is preferable to a lambda, as it's easier to reason about the structure of the code; and should the need arise, you can debug or modify the gtk handler independent of the `value-changed` signal handler.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
As I indicated in my previous review, passing flake8 is a minimum requirement.

**Comment by meow464** [comment]
Sorry I had noqa F821 on my config file. Would it be ok to add `# noqa F821` to those lines? Flake8's complain is that I used a string as a type annotation (I didn't need the class for anything else so it didn't make sense to import it).

**Comment by freakboy3742** [comment]
PEP484 moved out of provisional status in Py3.8; AIUI, that means string-form annontations are only legal as forward references, which makes the flake8 warning accurate. 

Is there any reason we can't just import the required classes and make the type annotations concrete? If that won't work - are the type annotations actually doing anything other than documentation?

**Comment by meow464** [comment]
Type annotations are just documentation in this case. I didn't want to import the class because it's not used for anything else. I will move the annotations to a comment.

**Comment by meow464** [comment]
A test failed. This didn't show up until some hours after I requested the review.

**Comment by freakboy3742** [comment]
Unfortunately, because cryptocurrency enthusiasts destroy everything good in the world, Github has an "approve test run" workflow step on every pull request from a "first time contributor". I hit the approve button as soon as I saw the PR this morning.

That said - as with flake8 checks, the test suite can be run on your local machine. There's details in the Toga contribution docs.

**Comment by meow464** [comment]
Note: The bot says coverage decreases for `src/core/toga/window.py` but I didn't change this file.

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 24]
```diff
@@ -1,92 +1,231 @@
...
     def insert(self, index, item):
-        self.store.insert(item)
+        """
+        Args:
+            index (int)
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+
+        if index == -1:
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:111]: Why is this special case required?

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 25]
```diff
@@ -1,92 +1,231 @@
...
+        Args:
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+        index = item._impl.get_index()
+
+        if index == -1:
+            self.store.append(new_item)
+        else:
+            self.store.insert(index, new_item)
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:130]: Won't this result in every row change adding a new row to the list?

File: src/gtk/toga_gtk/widgets/detailedlist.py [hunk 26]
```diff
@@ -1,92 +1,231 @@
...
 
     def remove(self, item, index):
-        self.store.remove(item, index)
+        """
+        Removes a row from the store. This method doesn't trigger the `on_delete` callback and it
+        doesn't remove the row from the interface.
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/detailedlist.py:135]: err... it doesn't? Why not? What removes the row from the interface?

File: src/gtk/toga_gtk/widgets/internal/rows/texticon.py [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+    @property
+    def subtitle(self):
+        return self.interface.subtitle
+
+    def get_icon(self, row, factory):
+        if getattr(row, "icon") is None:
+            return None
+        else:
+            row.icon.bind(factory)
+            # TODO: see get_scale_factor() to choose 72 px on hidpi
```
Comment (freakboy3742) [inline src/gtk/toga_gtk/widgets/internal/rows/texticon.py:59]: Flagging this TODO to be TODONE :-)

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I've flagged a couple more things to investigate/clarify/clean up; unfortunately I've been ill this weekend and haven't got as much done as I was hoping. I want to do another pass of hands-on testing, but I think we're getting close to being ready to merge.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 24 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:111]: That was a mistake. In Toga -1 means last item but on Gtk it means not present. I thought -1 meant it was the last item in the list and that `insert` didn't accept -1, thus I used `append`.
Calling `row_factory()` before `item._impl.get_index()` was replacing `item._impl` with a new `Gtk.ListBoxRow` which of course was not yet in the list. I fixed it and changed some variables names' to avoid future confusion.

Also changed `STYLES`to `TOGA_DEFAULT_STYLES`.

Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:111]: Hmm... looking at [the docs](https://web.archive.org/web/20210410062839/https://developer.gnome.org/gtk3/stable/GtkListStore.html#gtk-list-store-insert) -1 should be a valid index for `insert()`:

>  If position is -1 or is larger than the number of rows on the list, then the new row will be appended to the list.

But `insert(-1, item)` gives: `OverflowError: -1 not in range 0 to 4294967295`. This might be a bug on Gtk, but in any case I also made the mistake of calling `row_factory()` before `item._impl.get_index()`.

I'm using archive.org because the live page is currently giving me a 404. Since I'm on Debian stable the archive should be recent enough.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 25 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:130]: Yes, this must have happened during the above confusion. I added this case to `test_change()`.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 24 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:111]: Oh I was reading the docs for `Gtk.ListStore` not `Gio.ListStore`.

File: src/gtk/toga_gtk/widgets/detailedlist.py (hunk 26 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/detailedlist.py:135]: The user, by calling `dl.data.remove()`, which in turn calls `dl_gtk.remove()`. The remove function should call `on_delete()` though.

This brought another problem to my attention: When a row is removed gtk immediately calls `dl_gtk.gtk_on_row_selected()`, and thus the `dl.on_select()` callback.
On the detailedlist example, at the bottom of the window, it briefly displays "No row selected" and then "Row English is going to be deleted.". This is probably always going to work because signals are very fast, but signals are also asynchronous so there is no guarantee.

File: src/gtk/toga_gtk/widgets/internal/rows/texticon.py (hunk 27 above)
Comment (meow464) [reply src/gtk/toga_gtk/widgets/internal/rows/texticon.py:59]: I did as it's done [here](https://github.com/beeware/toga/blob/19e89a77f1bacddc09110fd5c240ea4ecbf85b2a/src/gtk/toga_gtk/widgets/imageview.py#L30) but I don't have a hidpi screen to test.

**Comment by meow464** [comment]
I hope you get better!

**Review by freakboy3742 — APPROVED** [review]
Apologies for the delay on the final review. I've pushed a couple of minor code style cleanups; but otherwise, this looks like it's good to go!
